### ISSUE TITLE
libgit2 1.8.2 regression - git_remote_fetch causes crash

### ISSUE THREAD

**Issue by sans-c** [libgit2/libgit2#6930]
Description below - if you need any more info, please let me know.

### Reproduction steps

This issue popped up when using Siril. Try to start siril-git with libgit2 1.8.2 installed and it will crash. The Siril issue was filed here: https://gitlab.com/free-astro/siril/-/issues/1407 and provides further information and a backtrace which points to libgit2.

### Expected behavior

Downgrading to libgit2 1.8.1 makes the issue disappear.

### Actual behavior

Siril crashes with libgit2 1.8.2

### Version of libgit2 (release number or SHA1)

1.8.2

### Operating system(s) tested

Endeavour/Arch

**Comment by ethomson** [comment]
That's pretty surprising; there are basically no functional changes between v1.8.1 and v1.8.2. 🤔 

We have changed http-parser, which is implicated in the stack trace, but that won't land in a release until v1.9...

Is this really v1.8.2 without changes? Just confirming, what's the commit sha?

**Comment by ervinpopescu** [comment]
@sans-c looks like this also comes up when using `git_clone` -> `clone_into` -> `git_remote_connect_ext` so I'd say the title is "wrong", could you please correct that?

@ethomson for me on arch version is 1.8.2 from the github archive (releases): https://gitlab.archlinux.org/archlinux/packaging/packages/libgit2/-/blob/main/PKGBUILD?ref_type=heads#L30.

tar is https://github.com/libgit2/libgit2/archive/v1.8.2/libgit2-v1.8.2.tar.gz.

Sooo, `4516ca149` with actually relevant last commit `4516ca149`?

**Comment by ethomson** [comment]
I was wrong; http-parser changes actually landed in v1.8.1. So there’s probably something busted in there; this is probably related to the change we made to allow users to opt in to llhttp. But I’m surprised that this wasn’t a problem in v1.8.1.

**Comment by ervinpopescu** [comment]
```
#11 git_remote_fetch (remote=0x51300003bd80, refspecs=0x0, opts=0x7fffdc309050, reflog_message=0x5555563b60e0 "fetch")
    at /usr/src/debug/libgit2/libgit2-1.8.2/src/libgit2/remote.c:1375
```
Ah, I was just about to ask if `refspecs` value here was okay. 

> this is probably related to the change we made to allow users to opt in to llhttp.

Could you please share the commit sha for that?

> But I’m surprised that this wasn’t a problem in v1.8.1.

So that probably means it's still something between v1.8.1 and v1.8.2 which triggered it 😄 gotta love bugs.

**Comment by ethomson** [comment]
Something in https://github.com/libgit2/libgit2/pull/6713 is the culprit, I’m not sure if it’s that hash precisely; it looks like we didn’t communicate the build flag changes very well, so the arch build for v1.8.1 used our bundled http parser. So that’s what we didn’t see this in the v1.8.1 build, I think.

**Comment by ethomson** [comment]
We might need to backport https://github.com/libgit2/libgit2/pull/6870 and create a v1.8.3

**Comment by ervinpopescu** [comment]
> We might need to backport #6870 and create a v1.8.3

Yes!

Segfault (for me, but also probably for everyone) occurrs when invoking `CALLBACK_DATA(status)` at line 910.

Testing with commit e3597e9ff:

```bash
$ ldd /usr/lib/libgit2.so.1.8.2
        linux-vdso.so.1 (0x000078d804f33000)
        libssl.so.3 => /usr/lib/libssl.so.3 (0x000078d804ce2000)
        libcrypto.so.3 => /usr/lib/libcrypto.so.3 (0x000078d804600000)
        libllhttp.so.9.2 => /usr/lib/libllhttp.so.9.2 (0x000078d804cd0000)
        libpcre2-8.so.0 => /usr/lib/libpcre2-8.so.0 (0x000078d804c31000)
        libssh2.so.1 => /usr/lib/libssh2.so.1 (0x000078d804be8000)
        libz.so.1 => /usr/lib/libz.so.1 (0x000078d804bcf000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x000078d804b9f000)
        libc.so.6 => /usr/lib/libc.so.6 (0x000078d80440f000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x000078d804f35000)
$ git2-aur # reproducer
works
```

Switching to llhttp also worked:
```diff
diff --git i/PKGBUILD w/PKGBUILD
index 6024335..4f7e789 100644
--- i/PKGBUILD
+++ w/PKGBUILD
@@ -16,7 +16,7 @@ license=(LicenseRef-GPL-2.0-only-with-linking-exception)
 depends=(
   gcc-libs
   glibc
-  http-parser
+  llhttp
   zlib
 )
 makedepends=(
@@ -38,7 +38,7 @@ build() {
     -D CMAKE_INSTALL_PREFIX=/usr
     -D REGEX_BACKEND=pcre2
     # NOTE: may require further consolidation: https://github.com/libgit2/libgit2/issues/6923
-    -D USE_HTTP_PARSER=http-parser
+    -D USE_HTTP_PARSER=llhttp
     -D USE_SSH=ON
     -S $pkgname-$pkgver
     -W no-dev
```

```bash
$ ldd /usr/lib/libgit2.so.1.8.2
        linux-vdso.so.1 (0x000074d030149000)
        libssl.so.3 => /usr/lib/libssl.so.3 (0x000074d02fee9000)
        libcrypto.so.3 => /usr/lib/libcrypto.so.3 (0x000074d02f800000)
        libllhttp.so.9.2 => /usr/lib/libllhttp.so.9.2 (0x000074d02fed7000)
        libpcre2-8.so.0 => /usr/lib/libpcre2-8.so.0 (0x000074d02fe38000)
        libssh2.so.1 => /usr/lib/libssh2.so.1 (0x000074d02fdef000)
        libz.so.1 => /usr/lib/libz.so.1 (0x000074d02fdd6000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x000074d02fda6000)
        libc.so.6 => /usr/lib/libc.so.6 (0x000074d02f60f000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x000074d03014b000)
$ git2-aur # reproducer
works
```
I hope I understood correctly how to enable llhttp :)

**Comment by ervinpopescu** [comment]
> Something in #6713 is the culprit, I’m not sure if it’s that hash precisely; it looks like we didn’t communicate the build flag changes very well, so the arch build for v1.8.1 used our bundled http parser. So that’s what we didn’t see this in the v1.8.1 build, I think.

I don't have a way to check that... Unless you can tell me which version it was, maybe I could see it in `ldd`?

**Comment by ethomson** [comment]
@ervinpopescu Thanks, that's really helpful. Would you be so kind as to check https://github.com/libgit2/libgit2/pull/6931? If that's 👍 then I'll cut a v1.8.3 at that commit.

**Comment by ervinpopescu** [comment]
```
$ ldd /usr/lib/libgit2.so.1.8.2
        linux-vdso.so.1 (0x00007402a9ba8000)
        libssl.so.3 => /usr/lib/libssl.so.3 (0x00007402a9948000)
        libcrypto.so.3 => /usr/lib/libcrypto.so.3 (0x00007402a9400000)
        libhttp_parser.so.2.9 => /usr/lib/libhttp_parser.so.2.9 (0x00007402a993e000)
        libpcre2-8.so.0 => /usr/lib/libpcre2-8.so.0 (0x00007402a9361000)
        libssh2.so.1 => /usr/lib/libssh2.so.1 (0x00007402a98f5000)
        libz.so.1 => /usr/lib/libz.so.1 (0x00007402a9348000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007402a931a000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007402a9129000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007402a9baa000)

$ rm qtile -rf && cargo build -j $(nproc) -q && git2-aur
works
```
🎉

Feel free to release!

**Comment by ethomson** [comment]
Awesome; thanks everybody for the assistance. 🙏

**Comment by ethomson** [comment]
Fixed via #6931
